### ISSUE TITLE
[Bug] xattention在ascendc后端精度没问题，pto后端精度最大diff=0.045

### ISSUE THREAD

**Issue by liuwenda4** [tile-ai/tilelang-ascend#1273]
## Bug Description
https://github.com/tile-ai/tilelang-ascend/pull/1255 这是xattention的相关代码的pr，里面包含有算子信息以及临时接口信息
shared_O_merge_ws 和 shared_gl_merge_ws 的 rescale 路径仍有差异。shared_gm_merge_ws（max，不 rescale）一致，说明 rescale 操作（TMUL/TROWEXPANDMUL 乘以 exp(old_max - new_max)）有问题
## Environment

**Hardware:**
- NPU Device: [e.g. Ascend A2, A3]
- Other hardware info: [e.g. CPU, memory]
A3
**Software:**
- OS: [e.g. Ubuntu 20.04, EulerOS]
- CANN version: [e.g. 8.3.RC1]
- torch-npu version: [e.g. 2.6.0.RC1]
- Python version: [e.g. 3.10]
- TileLang-Ascend version: [e.g. commit hash or tag]
- Installation method: [wheel package / build from source]

## Code to Reproduce

```python
# Minimal reproducible code
import tilelang
import torch

# Your code here
```
https://github.com/tile-ai/tilelang-ascend/pull/1255
## Error Message

```
Paste the full error message here

```

## Expected Behavior
pto后端精度通过
## Actual Behavior
A clear and concise description of what actually happened.
ascendc后端精度通过，pto后端精度不通过
## Additional Context

**Programming Mode:**
- [ ] Developer mode (automatic sync/buffer management)
- [x] Expert mode (manual T.Scope/T.barrier_all)
- [ ] Mixed mode

**Backend:**
- [x] Ascend C & PTO backend
- [ ] AscendNPU IR backend

**Pass Configs (if applicable):**
```python
pass_configs = {
    tilelang.PassConfigKey.TL_ASCEND_AUTO_SYNC: ...,
    tilelang.PassConfigKey.TL_ASCEND_MEMORY_PLANNING: ...,
    # ...
}
```

**Operator Type:**
[e.g. GEMM, Flash Attention, Softmax, Convolution, etc.]
Flash Attention
**Additional Information:**
- Have you checked the [Programming Guide](../docs/TileLang-Ascend%20Programming%20Guide.md)?
- Have you searched existing issues?
- Any other context or screenshots?

**Comment by ArmandAlbert** [comment]
Please check https://gitcode.com/cann/pto-isa/tree/master/tests/npu/a2a3/src/st/testcase/trowexpandsub and review the api implement.
If there are differences in usage or use case coverage compared to the pto-isa use case coverage, modifications can be attempted.

**Comment by platelett** [comment]
## Resolution

根因是 **PTO 后端的 V-pipe write-read hazard**：当一条 V-pipe 二元指令（如 `TSUB`）写入 UB buffer 后，紧跟的下一条 V-pipe 指令（如 `TEXP`）读取同一 buffer 时，写入尚未刷到 UB 内存，读到了 stale data。

具体到 xattention，softmax rescale 路径中：

```python
T.tile.sub(m_i_prev, m_i_prev, m_i)    # TSUB: m_i_prev = m_i_prev - m_i
T.tile.exp(m_i_prev, m_i_prev)          # TEXP: 立即读 m_i_prev → 读到旧值
```

PTO 后端生成 `vsub; vexp` 背靠背，中间零条指令，`vexp` 在 `vsub` 写入落地前就发射了。AscendC 后端不受影响，是因为它在每条向量指令前后都插入了 mask 设置/恢复的标量指令（`set_mask_norm / set_vector_mask` 等），这些 S-pipe 指令的插入触发了某种硬件的保护机制，确保前一条 V-pipe 写入完成后才允许后续 V-pipe 读取同一 buffer。

### 修复

在所有存在 V-pipe 写后读依赖的位置插入 `T.pipe_barrier("v")`，确保 V-pipe drain。已通过修改 `examples/xattention/xattention.py` 和 `xattention_paged.py` 解决。

### 说明

这不是编译器 bug。当 `pass_configs` 中 `TL_ASCEND_AUTO_SYNC=False`（Expert 模式）时，`pipe_barrier` 的插入是用户的责任。`AUTO_SYNC=True` 时编译器会自动处理。

但这个行为确实是一个常见陷阱：
- AscendC 后端天然不受影响（mask 指令触发硬件 interlock），PTO 后端则必须显式加 barrier
- 同一份 TileLang 代码在两个后端表现不一致，容易误导开发者
- `pipe_barrier("v")` 只有 PTO 后端才真正需要

**建议**：将 V-pipe barrier 的自动插入从 `TL_ASCEND_AUTO_SYNC` 中分离为独立的 pass_config 选项，使得 Expert 模式下也能选择性地开启这一项自动保护，避免反复踩坑。

**Comment by ArmandAlbert** [comment]
On A3 machine, pipe_v can be added between vector instructions with buffer data dependencies to avoid issues with the pto backend; this problem does not exist on the A5.

**Comment by 2Sheldonlaw** [comment]
Fixed by #1314（3b30ff04，2026-07-09 合入）。

根因不是 rescale（TMUL/TROWEXPANDMUL），而是 L1→L0B 越界：CopyL1ToL0Codegen 对切片 3D L1 buffer（shared_v_stack_l1 [4,128,128]）误用物理行数 512，FindBestTileRowB(512,128)→256 > L0B 容量 128 → V 操作数污染 → shared_O/gl_merge_ws 损坏；shared_gm_merge_ws 不走该路径，故两后端一致——与您观察吻合。#1314 修正为 slice_valid_row（256）→ tile_row=128。

您的复现树 = ce07b80c（07-03）+ cherry-pick #1283（07-15），恰好缺 #1314（07-09）。使用 #1283 合入后的任意分支即已包含。

最新分支验证（您的原始步骤）：non-paged / paged 均 0/524288 mismatch（原 91.3%），max abs diff 0.000488 < 0.001；PTO vs ascendc 同输入输出逐位一致。请更新复测，可关闭。

**Comment by 2Sheldonlaw** [comment]
[https://github.com/tile-ai/tilelang-ascend/pull/1314](url)
[https://github.com/tile-ai/tilelang-ascend/pull/1283](url)
这两个合入的PR已经修复issue的问题，应该是可以关闭目前的issue
